score.go

```diff
@@ -0,0 +1,952 @@
+	}
+
+	// check P3b
+	if p.MeshFailurePenaltyWeight > 0 {
+		return fmt.Errorf("invalid MeshFailurePenaltyWeight; must be negative (or 0 to disable)")
+	}
+	if p.MeshFailurePenaltyWeight != 0 && (p.MeshFailurePenaltyDecay <= 0 || p.MeshFailurePenaltyDecay >= 1) {
+		return fmt.Errorf("invalid MeshFailurePenaltyDecay; must be between 0 and 1")
+	}
+
+	// check P4
+	if p.InvalidMessageDeliveriesWeight >= 0 {
+		return fmt.Errorf("invalid InvalidMessageDeliveriesWeight; must be negative")
+	}
+	if p.InvalidMessageDeliveriesDecay <= 0 || p.InvalidMessageDeliveriesDecay >= 1 {
+		return fmt.Errorf("invalid InvalidMessageDeliveriesDecay; must be between 0 and 1")
+	}
+
+	return nil
+}
+
+// router interface
+func (ps *peerScore) Start(gs *GossipSubRouter) {
+	if ps == nil {
+		return
+	}
+
+	ps.msgID = gs.p.msgID
+	ps.host = gs.p.host
+	go ps.background(gs.p.ctx)
+}
+
+func (ps *peerScore) Score(p peer.ID) float64 {
+	if ps == nil {
+		return 0
+	}
+
+	ps.Lock()
+	defer ps.Unlock()
+
+	return ps.score(p)
+}
+
+func (ps *peerScore) score(p peer.ID) float64 {
+	pstats, ok := ps.peerStats[p]
+	if !ok {
+		return 0
+	}
+
+	var score float64
+
+	// topic scores
+	for topic, tstats := range pstats.topics {
+		// the topic score
+		var topicScore float64
+
+		// the topic parameters
+		topicParams, ok := ps.params.Topics[topic]
+		if !ok {
+			// we are not scoring this topic
+			continue
+		}
+
+		// P1: time in Mesh
+		if tstats.inMesh {
+			p1 := float64(tstats.meshTime / topicParams.TimeInMeshQuantum)
+			if p1 > topicParams.TimeInMeshCap {
+				p1 = topicParams.TimeInMeshCap
+			}
+			topicScore += p1 * topicParams.TimeInMeshWeight
+		}
+
+		// P2: first message deliveries
+		p2 := tstats.firstMessageDeliveries
+		topicScore += p2 * topicParams.FirstMessageDeliveriesWeight
+
+		// P3: mesh message deliveries
+		if tstats.meshMessageDeliveriesActive {
+			if tstats.meshMessageDeliveries < topicParams.MeshMessageDeliveriesThreshold {
+				deficit := topicParams.MeshMessageDeliveriesThreshold - tstats.meshMessageDeliveries
+				p3 := deficit * deficit
+				topicScore += p3 * topicParams.MeshMessageDeliveriesWeight
+			}
+		}
+
+		// P3b:
+		p3b := tstats.meshFailurePenalty
+		topicScore += p3b * topicParams.MeshFailurePenaltyWeight
+
+		// P4: invalid messages
+		p4 := tstats.invalidMessageDeliveries
+		topicScore += p4 * topicParams.InvalidMessageDeliveriesWeight
+
+		// update score, mixing with topic weight
+		score += topicScore * topicParams.TopicWeight
+	}
+
+	// P5: application-specific score
+	p5 := ps.params.AppSpecificScore(p)
+	score += p5 * ps.params.AppSpecificWeight
+
+	// P6: IP collocation factor
+	for _, ip := range pstats.ips {
+		peersInIP := len(ps.peerIPs[ip])
+		if peersInIP > ps.params.IPColocationFactorThreshold {
+			surpluss := float64(peersInIP - ps.params.IPColocationFactorThreshold)
+			p6 := surpluss * surpluss
+			score += p6 * ps.params.IPColocationFactorWeight
```

Comment:
Comment: this weight is negative -- it is a penalty score.

Reply:
the negative comment exists in the parameters struct, it's redundant here.

Reply:
I'm asking for a code comment, not API docs. I'll add it myself. (We'll all be maintaining this code in the future.)